### ISSUE TITLE
Global tools should be allowed to update themselves

### ISSUE THREAD

**Issue by ardove** [dotnet/cli#11376]
I previously opened https://github.com/dotnet/cli/issues/11375 but on the wrong account, so reopening here.

This is handy for making sure that someone using your tool doesn't need to understand the dotnet cli to update their tools (we can simply provide useful contextual information from our global tool that guides them through the process). Typically, something like a `<toolName> update` is less verbose and easier to remember than `dotnet tool install -g <toolName>`). 

## Steps to reproduce
* Create a global tool which allows for an update command (`dotnet tool install -g <toolName>`)
* Publish version 0.0.0 of the tool
* Implement the update command in such a way that it invokes `dotnet tool update -g <toolName>` to update to the latest version of the tool
* Install version 0.0.0 of the tool
* Publish version 0.0.1 of the tool 
* Run the update command (`<toolName> update`)

## Expected  behavior
* The tool can update itself successfully
* Any future invocations of the tool use the updated version 

## Actual behavior
```
Failed to uninstall tool package '<toolName>': Access to the path '<userPath>\.dotnet\tools\.store\<toolName>\0.0.0' is denied
```

## Environment data
```
.NET Core SDK (reflecting any global.json):
 Version:   2.2.104
 Commit:    73f036d4ac

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.104\

Host (useful for support):
  Version: 2.2.2
  Commit:  a4fd7b2c84

.NET Core SDKs installed:
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.302 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.401 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.500 [C:\Program Files\dotnet\sdk]
  2.1.502 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.1.504 [C:\Program Files\dotnet\sdk]
  2.1.505 [C:\Program Files\dotnet\sdk]
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.2.104 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3-servicing-26724-03 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

**Comment by livarcocc** [comment]
@wli3

**Comment by wli3** [comment]
Thank you for the suggestion!

For UX perspective, `<toolName> update` is ambiguous with a command from the tool itself.

For technical side. Once a command is installed, SDK is out of the picture, and when invoking, it is just like run a console app. So, it would be hard to add extra SDK action in the middle. At the same time, this design makes tool producers' testing much easier and it improves start up time.

I think the root problem is the command being too long. It is a tradeoff between discoverability and easy to use. And at this point we are more on the discoverability side.

So, I lean forward not implement an extra command.

cc @KathleenDollard

**Comment by ardove** [comment]
I don't know that it is ambiguous, actually. A good example is the nuget command line interface. It has an option to update itself `nuget update --self`. The concept is the same (except we're just assuming the --self option). 

The problem is that the command installed isn't versioned so I can't install multiple versions side by side. In my case, it's installed under C:\Users\<userName>\.dotnet\tools as <toolName>.exe. 

I guess the workaround for me is to not use the global install path and manage the installed version manually. Rather than running a `dotnet tool update`, I'd need to run a `dotnet install tool <versionedToolPath/>` and then a `dotnet uninstall tool <previousVersionToolPath>` at some point later (forcing a user to restart for the action to take effect).

**Comment by wli3** [comment]
> The problem is that the command installed isn't versioned

Could you see if [local tools](https://github.com/dotnet/cli/issues/10288) can solve this problem? tool-version is scoped to a directory.

**Comment by ardove** [comment]
It's close, but we really want folks to be able to use our tool from wherever they are (not just inside a specific directory).

**Comment by wli3** [comment]
We had conversion about supporting multiple versions of the same tool. However, we cannot find a good UX experience -- we don't know how to invoke it. `dotnetsay --version 1.3` vs `dotnetsay --version 1.4` could be hard to type. Any suggestions?

> be able to use our tool from wherever they are

A command case we find is a project require to use a lower version, so a local tool manifest file could help pin the version. Does your scenario require to run multiple versions on a same project/files?

**Comment by ardove** [comment]
In fact, our case is the opposite. We want folks to always use the latest version of our tool (and it’s dependencies). 

For a concrete example, we have a number of dotnet templates that need to be invoked to get a complete solution. This is by design as some folks will want to pick and choose which templates they use, but the majority of cases will just want “the whole enchilada”. So we wrapped up the invocations of several dotnet cli commands into a global tool so that it’s a one liner instead of a 5 liner. Updating those templates in process is easy. Updating the tool which offers the wrapped up commands is what we’re after.

The suggestion I offered about allowing multiple versions to be installed doesn’t mean that I’d expect multiple versions should be active or even invocable. It just means that multiple versions of the executable could be found in the dotnet tool global install locations, but the tool’s manifest would only ever point at one of them. This is consistent with the way that dotnet template packages are installed (along with every other nuget package in the users local nuget cache). Doing this would allow the dotnet tool update command to just install the next version in the dotnet tool directory and just update the tools manifest so that it now points to the new version.

**Comment by KathleenDollard** [comment]
Sounds like there are two suggestions here:

1) Support `<toolName> update`
2) Support multiple versions

I'm not clear on your reasons for 2) other than supporting 1) so I'll focus on that and let you follow up on what I may be missing. 

While will pointed out some good supporting reasons, the reason we just can't support `<toolName> update` is that it would be breaking change. We must assume someone somewhere has created an exe that takes `update` as an argument. It's not just global tools, since we supported the `dotnet-` syntax prior to the global tool creation. So unfortunately, this particular approach is not even an option.

Could your tool shell out to `dotnet tool update <toolName>`?

**Comment by ardove** [comment]
We attempted starting a new process that invoked `dotnet tool update <toolName>`. The problem is that there's a lock on the `<toolName>` executable because it's currently in use which is why that doesn't work (hence, my suggestion that the dotnet cli allow a caller to install multiple versions side by side).

I'm not sure I understand why it would be a breaking change. The `dotnet-` syntax could still work. The fix I'm suggesting would just mean multiple executables would be  installed under `C:\Users\<userName>.dotnet\tools` and that the dotnet cli would need to know which one to pick (based on some metadata that lives on disk... NOT by specifying the version when invoking).

In any case, I think we've belabored the point enough and I'm not convinced that adding this capability to the .net cli would be helpful to anyone other than us, so we'll find a way around it. Thanks for the discussion and insight!

Feel free to close this at your convenience.
